packages/sitecore-jss-nextjs/src/editing/feaas-render-middleware.ts

```diff
@@ -0,0 +1,152 @@
+   */
+  constructor(protected config?: FEAASRenderMiddlewareConfig) {
+    super(config);
+
+    this.resolvePageUrl = config?.resolvePageUrl ?? this.defaultResolvePageUrl;
+  }
+
+  /**
+   * Gets the Next.js API route handler
+   * @returns route handler
+   */
+  public getHandler(): (req: NextApiRequest, res: NextApiResponse) => Promise<void> {
+    return this.handler;
+  }
+
+  private handler = async (req: NextApiRequest, res: NextApiResponse): Promise<void> => {
+    const { method, query, body, headers } = req;
+
+    const startTimestamp = Date.now();
+
+    debug.editing('feaas render middleware start: %o', {
+      method,
+      query,
+      headers,
+      body,
+    });
+
+    if (method !== 'GET') {
+      debug.editing('invalid method - sent %s expected GET', method);
+      res.setHeader('Allow', 'GET');
+      return res.status(405).json({
+        html: `<html><body>Invalid request method '${method}'</body></html>`,
+      });
+    }
+
+    // Validate secret
+    const secret = query[QUERY_PARAM_EDITING_SECRET];
+    if (secret !== getJssEditingSecret()) {
+      debug.editing(
+        'invalid editing secret - sent "%s" expected "%s"',
+        secret,
+        getJssEditingSecret()
+      );
+      return res.status(401).json({
+        html: '<html><body>Missing or invalid secret</body></html>',
+      });
+    }
+
+    try {
+      // Resolve server URL
+      const serverUrl = this.resolveServerUrl(req);
+
+      // Get query string parameters to propagate on subsequent requests (e.g. for deployment protection bypass)
+      const params = this.getQueryParamsForPropagation(query);
+
+      // Enable Next.js Preview Mode
+      res.setPreviewData({});
+
+      // Grab the Next.js preview cookies to send on to the render request
+      const cookies = res.getHeader('Set-Cookie') as string[];
+
+      // Make actual render request for page route, passing on preview cookies as well as any approved query string parameters.
+      // Note timestamp effectively disables caching the request in Axios (no amount of cache headers seemed to do it)
+      const requestUrl = new URL(this.resolvePageUrl(serverUrl));
+
+      for (const key in params) {
+        if ({}.hasOwnProperty.call(params, key)) {
+          requestUrl.searchParams.append(key, params[key]);
+        }
+      }
+
+      // Pass "feaasSrc" in case a FEAASComponent is being requested
+      if (query.feaasSrc) {
+        requestUrl.searchParams.append('feaasSrc', query.feaasSrc as string);
+      }
+
+      requestUrl.searchParams.append('timestamp', Date.now().toString());
+
+      debug.editing('fetching page route for %s', requestUrl.toString());
+
+      const pageRes = await this.dataFetcher.get<string>(requestUrl.toString(), {
+        headers: {
+          Cookie: cookies.join(';'),
+        },
+      });
+
+      const html = pageRes.data;
+
+      if (!html || html.length === 0) {
+        throw new Error('Failed to render html');
+      }
```

Comment:
Wondering if we need to take the SSR approach with this one. The reasons we need to for Pages/EE integration (specific JSON format response expected, partial rendering) don't seem to apply here. Couldn't we just as easily redirect to the `/feaas/render` route with needed query string parms?

Reply:
It's a totally right call, it helped me to simplify the implementation